vespalib/src/tests/datastore/fixed_size_hash_map/fixed_size_hash_map_test.cpp

```diff
@@ -0,0 +1,312 @@
+// Copyright Verizon Media. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+
+#include <vespa/vespalib/datastore/fixed_size_hash_map.h>
+#include <vespa/vespalib/datastore/unique_store_allocator.h>
+#include <vespa/vespalib/datastore/unique_store_comparator.h>
+#include <vespa/vespalib/stllike/hash_map.h>
+#include <vespa/vespalib/util/lambdatask.h>
+#include <vespa/vespalib/util/rand48.h>
+#include <vespa/vespalib/util/size_literals.h>
+#include <vespa/vespalib/util/threadstackexecutor.h>
+#include <vespa/vespalib/gtest/gtest.h>
+
+#include <vespa/vespalib/datastore/unique_store_allocator.hpp>
+#include <vespa/vespalib/stllike/hash_map.hpp>
+
+#include <vespa/log/log.h>
+LOG_SETUP("vespalib_fixed_size_hash_map_test");
+
+using vespalib::datastore::EntryRef;
+using RefT = vespalib::datastore::EntryRefT<22>;
+using MyAllocator = vespalib::datastore::UniqueStoreAllocator<uint32_t, RefT>;
+using MyDataStore = vespalib::datastore::DataStoreT<RefT>;
+using MyCompare = vespalib::datastore::UniqueStoreComparator<uint32_t, RefT>;
+using GenerationHandler = vespalib::GenerationHandler;
+using vespalib::makeLambdaTask;
+using vespalib::GenerationHolder;
+using vespalib::datastore::FixedSizeHashMap;
+
+class FixedSizeHashMapHeld : public vespalib::GenerationHeldBase
+{
+    std::unique_ptr<const FixedSizeHashMap> _data;
+public:
+    FixedSizeHashMapHeld(size_t size, std::unique_ptr<const FixedSizeHashMap> data);
+    ~FixedSizeHashMapHeld();
+};
+
+FixedSizeHashMapHeld::FixedSizeHashMapHeld(size_t size, std::unique_ptr<const FixedSizeHashMap> data)
+    : GenerationHeldBase(size),
+      _data(std::move(data))
+{
+}
+
+FixedSizeHashMapHeld::~FixedSizeHashMapHeld() = default;
+
+struct DataStoreFixedSizeHashTest : public ::testing::Test
```

Comment:
Just copying the tests from simple_hash_map_test.cpp (with slight modifications) doesn't give any extra value in my opinion. The tests for FixedSizeHashMap should be on a lower level. No need for tests using multiple writers and readers that tests with a lot of data. That is already handled in simple_hash_map_test.cpp. We should write test cases the triggers the particular edge cases instead. For instance ensuring that the linked list is updated correctly when removing and ensuring that free nodes are re-used etc (perhaps we need to expose _free_count for this). I think only smoke_test and lookups_works_after_insert_and_remove is on the right abstraction level, the rest can go. Growing the hash map on the fly seems to just complicate things as well.

Reply:
Threaded tests have been removed.